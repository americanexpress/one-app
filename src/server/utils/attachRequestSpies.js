/*
 * Copyright 2023 American Express Travel Related Services Company, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
 * or implied. See the License for the specific language governing
 * permissions and limitations under the License.
 */

import https from 'node:https';
import http from 'node:http';
import url from 'node:url';
import onFinished from 'on-finished';

export function attachSpy(obj, methodName, spy) {
  const origMethod = obj[methodName];

  if (typeof origMethod !== 'function') {
    throw new TypeError(`${methodName} is not a function`);
  }

  if (typeof spy !== 'function') {
    throw new TypeError(`spy must be a function (was "${typeof spy}")`);
  }

  // we're monkeypatching, we need to reassign a property of the obj argument
  // eslint-disable-next-line no-param-reassign -- see above comment
  obj[methodName] = function spiedMethod(...args) {
    let returnValue;
    let originalCalled = false;
    const callOriginal = () => {
      originalCalled = true;
      returnValue = origMethod.apply(obj, args);
      return returnValue;
    };

    spy([...args], callOriginal);

    if (!originalCalled) {
      callOriginal();
    }

    return returnValue;
  };
}

function buildUrlObject(options, defaultProtocol) {
  // TODO: url.parse is deprecated, use new URL() instead
  // this is not a 1:1 replacement and will require changes.
  // Currently the parsed url is used to provide href to
  // loggers outgoingRequestEndSpy function.
  // https://github.com/americanexpress/one-app/blob/main/src/server/utils/logging/setup.js#L42
  // https://nodejs.org/api/url.html#urlparseurlstring-parsequerystring-slashesdenotehost
  const parsedPath = url.parse(options.path || '/');
  const protocol = options.protocol || `${defaultProtocol}:`;
  const urlObject = {
    auth: options.auth,
    hostname: options.hostname || options.host || 'localhost',
    protocol,
    port: options.port || (protocol === 'http:' ? 80 : 443),
    hash: parsedPath.hash,
    path: parsedPath.path,
    pathname: parsedPath.pathname,
    query: parsedPath.query,
    search: parsedPath.search,
  };
  if (
    (protocol === 'http:' && urlObject.port === 80)
    || (protocol === 'https:' && urlObject.port === 443)
  ) {
    urlObject.port = undefined;
  }
  return urlObject;
}

function parseUrl(options, defaultProtocol) {
  // some data is not stored in the clientRequest, have to duplicate some parsing
  // adapted from https://github.com/nodejs/node/blob/894203dae39c7f1f36fc6ba126bb5d782d79b744/lib/_http_client.js#L22
  if (typeof options === 'string') {
    return url.parse(options);
  }

  return url.parse(url.format(buildUrlObject(options, defaultProtocol)));
}

function httpSpy(defaultProtocol, requestSpy, socketCloseSpy) {
  return (args, callOriginal) => {
    const options = args[0];
    const clientRequest = callOriginal();
    const parsedUrl = parseUrl(options, defaultProtocol);

    requestSpy(clientRequest, parsedUrl);

    if (socketCloseSpy) {
      onFinished(clientRequest, () => socketCloseSpy(clientRequest, parsedUrl));
    }
  };
}

export default function attachRequestSpies(requestSpy, socketCloseSpy) {
  if (typeof requestSpy !== 'function') {
    throw new TypeError(`requestSpy must be a function (was "${typeof requestSpy}")`);
  }

  if (socketCloseSpy && (typeof socketCloseSpy !== 'function')) {
    throw new TypeError(
      `socketCloseSpy must be function if provided (was "${typeof socketCloseSpy}")`);
  }

  attachSpy(https, 'request', httpSpy('https', requestSpy, socketCloseSpy));
  attachSpy(http, 'request', httpSpy('http', requestSpy, socketCloseSpy));
}
