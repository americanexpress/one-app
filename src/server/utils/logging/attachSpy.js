/*
 * Copyright 2023 American Express Travel Related Services Company, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
 * or implied. See the License for the specific language governing
 * permissions and limitations under the License.
 */

export default function attachSpy(obj, methodName, spy) {
  const origMethod = obj[methodName];

  if (typeof origMethod !== 'function') {
    throw new TypeError(`${methodName} is not a function`);
  }

  if (typeof spy !== 'function') {
    throw new TypeError(`spy must be a function (was "${typeof spy}")`);
  }

  // we're monkeypatching, we need to reassign a property of the obj argument
  // eslint-disable-next-line no-param-reassign -- see above comment
  obj[methodName] = function monkeypatchedMethod(...args) {
    let returnValue;
    let originalCalled = false;
    const callOriginal = () => {
      originalCalled = true;
      returnValue = origMethod.apply(obj, args);
      return returnValue;
    };

    spy([...args], callOriginal);

    if (!originalCalled) {
      callOriginal();
    }

    return returnValue;
  };
}
