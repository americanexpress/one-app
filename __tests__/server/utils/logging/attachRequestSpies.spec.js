/*
 * Copyright 2023 American Express Travel Related Services Company, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except
 * in compliance with the License. You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software distributed under the License
 * is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND,either express
 * or implied. See the License for the specific language governing permissions and limitations
 * under the License.
 */

import http from 'node:http';
import https from 'node:https';
import onFinished from 'on-finished';

import attachRequestSpies from '../../../../src/server/utils/logging/attachRequestSpies';
import attachSpy from '../../../../src/server/utils/logging/attachSpy';

jest.mock('http', () => ({ request: jest.fn() }));
jest.mock('https', () => ({ request: jest.fn() }));
jest.mock('on-finished');
jest.mock('../../../../src/server/utils/logging/attachSpy');

describe('attachRequestSpies', () => {
  beforeEach(() => {
    attachSpy.mockClear();
  });

  it('throws if requestSpy is not a function', () => {
    expect(attachRequestSpies).toThrowErrorMatchingSnapshot();
  });

  it('does not throw if socketCloseSpy is not provided', () => {
    expect(() => attachRequestSpies(() => {})).not.toThrow();
  });

  it('throws if socketCloseSpy is provided but is not a function', () => {
    expect(() => attachRequestSpies(() => {}, 'apples')).toThrowErrorMatchingSnapshot();
  });

  it('attaches http and https spies', () => {
    attachRequestSpies(jest.fn());
    expect(attachSpy).toHaveBeenCalledTimes(2);
    expect(attachSpy.mock.calls[0][0]).toEqual(https);
    expect(attachSpy.mock.calls[0][1]).toBe('request');
    expect(attachSpy.mock.calls[1][0]).toEqual(http);
    expect(attachSpy.mock.calls[1][1]).toBe('request');
    expect(typeof attachSpy.mock.calls[0][2]).toBe('function');
    expect(typeof attachSpy.mock.calls[1][2]).toBe('function');
  });

  describe('requestSpy', () => {
    it('is called with clientRequest', () => {
      const requestSpy = jest.fn();
      attachRequestSpies(requestSpy);
      expect(attachSpy).toHaveBeenCalledTimes(2);

      const httpsSpy = attachSpy.mock.calls[0][2];
      const httpSpy = attachSpy.mock.calls[1][2];

      const callOriginal = jest.fn(() => 'client request object');

      httpsSpy(['http://example.tld'], callOriginal);
      httpSpy(['http://example.tld'], callOriginal);

      expect(requestSpy).toHaveBeenCalledTimes(2);
      expect(requestSpy.mock.calls[0][0]).toBe('client request object');
      expect(requestSpy.mock.calls[1][0]).toBe('client request object');
    });

    it('is called with object options', () => {
      const requestSpy = jest.fn();
      attachRequestSpies(requestSpy);
      expect(attachSpy).toHaveBeenCalledTimes(2);
      const httpsSpy = attachSpy.mock.calls[0][2];
      const httpSpy = attachSpy.mock.calls[1][2];
      const callOriginal = jest.fn(() => 'client request object');
      httpsSpy([{
        protocol: 'https',
        hostname: 'example.tld',
        port: 8080,
        method: 'GET',
        path: '/somewhere?over=rainbow#so-blue',
        auth: 'user:password',
      }], callOriginal);

      httpSpy([{
        protocol: 'http',
        hostname: 'example.tld',
        port: 8080,
        method: 'GET',
        path: '/somewhere?over=rainbow#so-blue',
        auth: 'user:password',
      }], callOriginal);

      expect(requestSpy).toHaveBeenCalledTimes(2);
      expect(requestSpy.mock.calls[0][1]).toMatchSnapshot();
      expect(requestSpy.mock.calls[1][1]).toMatchSnapshot();
    });

    it('is called with sparse object options', () => {
      const requestSpy = jest.fn();
      attachRequestSpies(requestSpy);
      expect(attachSpy).toHaveBeenCalledTimes(2);
      const httpsSpy = attachSpy.mock.calls[0][2];
      const httpSpy = attachSpy.mock.calls[1][2];

      const callOriginal = jest.fn(() => 'client request object');
      httpsSpy([{ method: 'GET' }], callOriginal);
      httpSpy([{ method: 'GET' }], callOriginal);
      expect(requestSpy).toHaveBeenCalledTimes(2);
      expect(requestSpy.mock.calls[0][1]).toMatchSnapshot();
      expect(requestSpy.mock.calls[1][1]).toMatchSnapshot();
    });

    it('is called with parsed options', () => {
      const requestSpy = jest.fn();
      attachRequestSpies(requestSpy);
      expect(attachSpy).toHaveBeenCalledTimes(2);
      const spy = attachSpy.mock.calls[0][2];
      const callOriginal = jest.fn(() => 'client request object');
      spy(['http://user:password@example.tld:8080/somewhere?over=rainbow#so-blue'], callOriginal);
      expect(requestSpy).toHaveBeenCalledTimes(1);
      expect(requestSpy.mock.calls[0][1]).toMatchSnapshot();
    });
  });

  describe('socketCloseSpy', () => {
    it('is called when the request socket closes', () => {
      onFinished.mockClear();
      const socketCloseSpy = jest.fn();
      attachRequestSpies(jest.fn(), socketCloseSpy);
      expect(attachSpy).toHaveBeenCalledTimes(2);
      const spy = attachSpy.mock.calls[0][2];
      const clientRequest = {};
      const callOriginal = () => clientRequest;
      spy(['http://example.tld'], callOriginal);
      expect(onFinished).toHaveBeenCalledTimes(1);
      onFinished.mock.calls[0][1]();
      expect(socketCloseSpy).toHaveBeenCalledTimes(1);
      expect(socketCloseSpy.mock.calls[0][0]).toBe(clientRequest);
      expect(socketCloseSpy.mock.calls[0][1]).toMatchSnapshot();
    });
  });
});
