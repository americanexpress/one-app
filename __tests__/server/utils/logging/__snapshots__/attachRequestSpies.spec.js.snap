// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`attachRequestSpies requestSpy is called with object options 1`] = `
Url {
  "auth": "user:password",
  "hash": "#so-blue",
  "host": "example.tld:8080",
  "hostname": "example.tld",
  "href": "https://user:password@example.tld:8080/somewhere?over=rainbow#so-blue",
  "path": "/somewhere?over=rainbow",
  "pathname": "/somewhere",
  "port": "8080",
  "protocol": "https:",
  "query": "over=rainbow",
  "search": "?over=rainbow",
  "slashes": true,
}
`;

exports[`attachRequestSpies requestSpy is called with object options 2`] = `
Url {
  "auth": "user:password",
  "hash": "#so-blue",
  "host": "example.tld:8080",
  "hostname": "example.tld",
  "href": "http://user:password@example.tld:8080/somewhere?over=rainbow#so-blue",
  "path": "/somewhere?over=rainbow",
  "pathname": "/somewhere",
  "port": "8080",
  "protocol": "http:",
  "query": "over=rainbow",
  "search": "?over=rainbow",
  "slashes": true,
}
`;

exports[`attachRequestSpies requestSpy is called with parsed options 1`] = `
Url {
  "auth": "user:password",
  "hash": "#so-blue",
  "host": "example.tld:8080",
  "hostname": "example.tld",
  "href": "http://user:password@example.tld:8080/somewhere?over=rainbow#so-blue",
  "path": "/somewhere?over=rainbow",
  "pathname": "/somewhere",
  "port": "8080",
  "protocol": "http:",
  "query": "over=rainbow",
  "search": "?over=rainbow",
  "slashes": true,
}
`;

exports[`attachRequestSpies requestSpy is called with sparse object options 1`] = `
Url {
  "auth": null,
  "hash": null,
  "host": "localhost",
  "hostname": "localhost",
  "href": "https://localhost/",
  "path": "/",
  "pathname": "/",
  "port": null,
  "protocol": "https:",
  "query": null,
  "search": null,
  "slashes": true,
}
`;

exports[`attachRequestSpies requestSpy is called with sparse object options 2`] = `
Url {
  "auth": null,
  "hash": null,
  "host": "localhost",
  "hostname": "localhost",
  "href": "http://localhost/",
  "path": "/",
  "pathname": "/",
  "port": null,
  "protocol": "http:",
  "query": null,
  "search": null,
  "slashes": true,
}
`;

exports[`attachRequestSpies socketCloseSpy is called when the request socket closes 1`] = `
Url {
  "auth": null,
  "hash": null,
  "host": "example.tld",
  "hostname": "example.tld",
  "href": "http://example.tld/",
  "path": "/",
  "pathname": "/",
  "port": null,
  "protocol": "http:",
  "query": null,
  "search": null,
  "slashes": true,
}
`;

exports[`attachRequestSpies throws if requestSpy is not a function 1`] = `"requestSpy must be a function (was "undefined")"`;

exports[`attachRequestSpies throws if socketCloseSpy is provided but is not a function 1`] = `"socketCloseSpy must be function if provided (was "string")"`;
